### ISSUE TITLE
parameter setting to reproduce the result in tstDemo.py

### ISSUE THREAD

**Issue by Lutieyuan** [hkaggarwal/modl#2]
Hi, I have set the parameter as your parer write: layers N =5, iterations K =10 , others are the same in  trn.py:
nLayers=5
epochs=100
batchSize=4
gradientMethod='MG'
K=10
sigma=0.01
both gradient method 'MG' and 'AG' have been tried, and variable reuse=tf.AUTO_REUSE also has been set. However, I can't reproduce the PSNR as high as 39dB in the provided model. My result is 33dB or less. 

So, I want to know, how could I set the  parameters in your code properly to reproduce your result?
Best !

**Comment by hkaggarwal** [comment]
Hi, 

To reproduce the results in the paper, please do the two-step training.
Step 1: Train a `K=1` model. It should give you PSNR in 31-33 dB range.
Step 2: Now train a `K=10` model by initializing with the weights of `K=1` model and PSNR should improve to 38-40 dB range. 

I would suggest using `batchSize=1` and `gradientMethod ='AG'` for both `K=1` and `K=10` training.

Let me know if the issue persists.

Bests,
Hemant

**Comment by Lutieyuan** [comment]
Hi!
Sorry for responding you late.

As the suggestions you provided, I use two training steps. The parameters of them are same to the recommendation.

After the k=1 per-trained model, the PSNR can reach 35~36 dB.
However,  problem occurs when training the K=10 model, which was initialized by the k=1 model.   In detail, I run the experiments two times. One of the experiments can reproduce the result (the per-trained k=1 model and the k=10 model are :'11May_0918pm_5L_1K_100E_AG' and '12May_0238pm_5L_10K_100E_AG'). But when I want to reproduce the result for the second time, things get wired. The per-trained model k=1 seems right (12May_0241pm_5L_1K_100E_AG) , but the result of  the k=10 model (13May_0204pm_5L_10K_100E_AG) drops to 15 dB. 
 
So is there probability that the second step training failed? Or there are some mistakes in my k=10 trn2_.py where I loaded the per-trained model for the continuing training ?

The following is the saved models of these two experiments and the codes for K=1 and K=10 model training.
[saveModels.zip](https://github.com/hkaggarwal/modl/files/3175631/saveModels.zip)

Thank you for your reply.
Best!

**Comment by hkaggarwal** [comment]
Hi,

Currently, I am traveling but early next week I will be able to give a close look to your savedModels.zip file.

Can you please try to train a K=2 model initialized with K=1 model and check if that gives slightly better results than the K=1 model.
Also, I would suggest giving a close look at the model created as well as the loss plot in Tensorborad output. The loss should be almost smoothly decaying in every epoch. 
Also, I would suggest to save the intermediate models (say every 10 epochs, just for debugging) and do the testing with them.

This is such a sharp decay in PSNR so certainly something is going wrong.

**Comment by Lutieyuan** [comment]
Hi!
I followed your suggestion and check the TrnLoss in the tensorboard. The following picture shows that the TrnLoss gets a strange boom at epoch 70 in the savedModel(13May_0204pm_5L_10K_100E_AG) 
![image](https://user-images.githubusercontent.com/35769977/58328083-c98f8900-7e63-11e9-874c-f7e68e2604ad.png).
But  I not sure what happens at  epoch 70.

**Comment by hkaggarwal** [comment]
I hope you save the models every 10 or 20 epochs. So if you have the model saved at any epoch less than 70, can you please try to test with that and let me know how it compares with K=1 model.
You might also want to check if this behavior (sudden jump) is consistent or just one time. Actually, I could not reproduce this issue.

**Comment by Lutieyuan** [comment]
HI, I did not saved the model every 10 or 20 epochs. But, from the TrnLoss, when the second training failed, the result of it may worse than the k=1 model. The following is the trnLoss and its corresponding result. The first is the k=1 model, the second and the third is the k=10 model, one is a succeed second training, and the other is a failed second training. Both of the second trainings are based on the same first training of the k=1 model.
![image](https://user-images.githubusercontent.com/35769977/58758528-f53efd00-854e-11e9-9502-06ceb0488f07.png)
![image](https://user-images.githubusercontent.com/35769977/58758532-1273cb80-854f-11e9-8732-5c49ac051aee.png)
![image](https://user-images.githubusercontent.com/35769977/58758535-1ef82400-854f-11e9-85a4-afa9df05d9b6.png)

In my reproducing, I am sure the failed second training may fail not consistently. But I not sure the code for the second training is built correctly. 
So could you check the trianing code I have uploaded before or if it is possible for you to upload your training code especially the training code for the second training which can reproduce the result in your paper?  If the same problem did not occur in your second training code, there must be some thing wrong with my second training code.

**Comment by hkaggarwal** [comment]
@Lutieyuan, Thanks for the detailed information. I have now updated the code such that we can easily initialize with a pre-trained MoDL with `K=1`. I checked that the code for `K=10` can reproduce the results in the paper. I have also added the latest trained model in the `savedModels` directory for testing purpose.

**Comment by Lutieyuan** [comment]
Hi, Thank you for your patient help. I believe this update will also help other people to reproduce your work correctly.
 So if one person want to use your network on another dataset and get a better result, he or she had better train a k=10 model initialized with the weights of a k=1 model?

**Comment by hkaggarwal** [comment]
@Lutieyuan, Yes that is true when you have fewer data and a small model architecture such as the simple 5-layer model that I am using. Initializing with pre-trained model helps in that case.

**Comment by Lutieyuan** [comment]
Ok, I see. I will try it. Thank you very much!
